src/bootstrap/native.rs

```diff
@@ -111,6 +116,276 @@ pub fn prebuilt_llvm_config(
+        // ORIGIN is a relative default, all binary and dynamic libraries we ship
+        // appear to have this (even when `../lib` is redundant).
+        "$ORIGIN/../lib:{}/lib",
+        t!(fs::canonicalize(nix_deps_dir)).to_str().unwrap()
+    );
+    patchelf.args(&["--set-rpath", &rpath_entries]);
+    if !fname.ends_with(".so") {
+        // Finally, set the corret .interp for binaries
+        let dynamic_linker_path = Path::new(nix_deps_dir).join("nix-support/dynamic-linker");
+        let dynamic_linker = t!(String::from_utf8(t!(fs::read(dynamic_linker_path))));
+        patchelf.args(&["--set-interpreter", dynamic_linker.trim_end()]);
+    }
+
+    builder.try_run(patchelf.arg(fname));
+}
+
+fn download_component(builder: &Builder<'_>, base: &str, url: &str, dest_path: &Path) {
+    // Use a temporary file in case we crash while downloading, to avoid a corrupt download in cache/.
+    let tempfile = t!(tempfile::NamedTempFile::new());
+    let temppath = tempfile.path().to_owned();
+    drop(tempfile);
+    let tempfile_str = temppath.to_str().expect("tempdir must be valid unicode");
+    // FIXME: support `do_verify` (only really needed for nightly rustfmt)
+    download_with_retries(builder, tempfile_str, &format!("{}/{}", base, url));
+    builder.rename(&temppath, dest_path);
+}
+
+fn download_with_retries(builder: &Builder<'_>, tempdir: &str, url: &str) {
+    println!("downloading {}", url);
+
+    // FIXME: check if curl is installed instead of skipping straight to powershell
+    if builder.build.build.contains("windows-msvc") {
+        for _ in 0..3 {
+            if builder.try_run(Command::new("PowerShell.exe").args(&[
+                "/nologo",
+                "-Command",
+                "[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12;",
+                &format!(
+                    "(New-Object System.Net.WebClient).DownloadFile('{}', '{}')",
+                    url, tempdir
+                ),
+            ])) {
+                return;
+            }
+            println!("\nspurious failure, trying again");
+        }
+    } else {
+        builder.run(Command::new("curl").args(&[
+            "-#",
+            "-y",
+            "30",
+            "-Y",
+            "10", // timeout if speed is < 10 bytes/sec for > 30 seconds
+            "--connect-timeout",
+            "30", // timeout if cannot connect within 30 seconds
+            "--retry",
+            "3",
+            "-Sf",
+            "-o",
+            tempdir,
+            url,
+        ]));
+    }
+}
+
+fn unpack(builder: &Builder<'_>, tarball: &Path, dst: &Path) {
+    println!("extracting {} to {}", tarball.display(), dst.display());
+    if !dst.exists() {
+        t!(fs::create_dir_all(dst));
+    }
+
+    // FIXME: will need to be a parameter once `download-rustc` is moved to rustbuild
+    const MATCH: &str = "rust-dev";
+    // `tarball` ends with `.tar.xz`; strip that suffix
```

Comment:
```suggestion
    // `tarball` ends with `.tar.xz`; strip .xz
```

Based on the uncompressed_filename below, this seems right?

Reply:
ah, so the code here is kind of misleading - here's the relevant bits:
```
    // `tarball` ends with `.tar.xz`; strip that suffix
    let uncompressed_filename =
        Path::new(tarball.file_name().expect("missing tarball filename")).file_stem().unwrap();

    // example: `rust-dev-nightly-x86_64-unknown-linux-gnu`
    let directory_prefix = Path::new(Path::new(uncompressed_filename).file_stem().unwrap());
```
I'll move the two lines closer together so it's more clear that the intermediate `.tar` path isn't used for anything.